File: persistence/src/main/java/io/sparkled/persistence/sequence/impl/SequencePersistenceServiceImpl.java
package io.sparkled.persistence.sequence.impl;

import io.sparkled.model.entity.Sequence;
import io.sparkled.model.entity.SequenceChannel;
import io.sparkled.model.entity.SongAudio;
import io.sparkled.model.entity.Stage;
import io.sparkled.model.render.RenderedStagePropDataMap;
import io.sparkled.persistence.QueryFactory;
import io.sparkled.persistence.sequence.SequencePersistenceService;
import io.sparkled.persistence.sequence.impl.query.*;

import javax.inject.Inject;
import java.util.*;

public class SequencePersistenceServiceImpl implements SequencePersistenceService {

    private QueryFactory queryFactory;

    @Inject
    public SequencePersistenceServiceImpl(QueryFactory queryFactory) {
        this.queryFactory = queryFactory;
    }

    @Override
    public Integer createSequence(Sequence sequence, byte[] songAudioData) {
        sequence = saveSequence(sequence, Collections.emptyList());

        SongAudio songAudio = new SongAudio().setSequenceId(sequence.getId()).setAudioData(songAudioData);
        return new SaveSongAudioQuery(songAudio).perform(queryFactory);
    }

    @Override
    public List<Sequence> getAllSequences() {
        return new GetAllSequencesQuery().perform(queryFactory);
    }

    @Override
    public Optional<Sequence> getSequenceById(int sequenceId) {
        return new GetSequenceByIdQuery(sequenceId).perform(queryFactory);
    }

    @Override
    public Optional<Stage> getStageBySequenceId(int sequenceId) {
        return new GetStageBySequenceIdQuery(sequenceId).perform(queryFactory);
    }

    @Override
    public Optional<SongAudio> getSongAudioBySequenceId(int sequenceId) {
        return new GetSongAudioBySequenceIdQuery(sequenceId).perform(queryFactory);
    }

    @Override
    public List<SequenceChannel> getSequenceChannelsBySequenceId(int sequenceId) {
        return new GetSequenceChannelsBySequenceIdQuery(sequenceId).perform(queryFactory);
    }

    @Override
    public Optional<SequenceChannel> getSequenceChannelByUuid(int sequenceId, UUID uuid) {
        return new GetSequenceChannelByUuidQuery(sequenceId, uuid).perform(queryFactory);
    }

    @Override
    public RenderedStagePropDataMap getRenderedStagePropsBySequence(Sequence sequence) {
        return new GetRenderedStagePropsBySequenceQuery(sequence).perform(queryFactory);
    }

    @Override
    public Map<String, UUID> getSequenceStagePropUuidMapBySequenceId(Integer sequenceId) {
        return new GetSequenceStagePropUuidMapBySequenceIdQuery(sequenceId).perform(queryFactory);
    }

    @Override
    public Sequence saveSequence(Sequence sequence, List<SequenceChannel> sequenceChannels) {
        sequence = new SaveSequenceQuery(sequence).perform(queryFactory);
        new SaveSequenceChannelsQuery(sequence, sequenceChannels).perform(queryFactory);
        return sequence;
    }

    @Override
    public void publishSequence(Sequence sequence, List<SequenceChannel> sequenceChannels, RenderedStagePropDataMap renderedStageProps) {
        sequence = new SaveSequenceQuery(sequence).perform(queryFactory);
        new SaveSequenceChannelsQuery(sequence, sequenceChannels).perform(queryFactory);
        new SaveRenderedStagePropsQuery(sequence, renderedStageProps).perform(queryFactory);
    }

    @Override
    public void deleteSequence(int sequenceId) {
        new DeleteSequencesQuery(Collections.singletonList(sequenceId)).perform(queryFactory);
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/sequence/impl/SequencePersistenceServiceImpl.kt
package io.sparkled.persistence.sequence.impl

import com.fasterxml.jackson.databind.ObjectMapper
import io.sparkled.model.entity.Sequence
import io.sparkled.model.entity.SequenceChannel
import io.sparkled.model.entity.Song
import io.sparkled.model.entity.SongAudio
import io.sparkled.model.entity.Stage
import io.sparkled.model.render.RenderedStagePropDataMap
import io.sparkled.persistence.QueryFactory
import io.sparkled.persistence.sequence.SequencePersistenceService
import io.sparkled.persistence.sequence.impl.query.DeleteSequencesQuery
import io.sparkled.persistence.sequence.impl.query.GetAllSequencesQuery
import io.sparkled.persistence.sequence.impl.query.GetRenderedStagePropsBySequenceQuery
import io.sparkled.persistence.sequence.impl.query.GetSequenceByIdQuery
import io.sparkled.persistence.sequence.impl.query.GetSequenceChannelByUuidQuery
import io.sparkled.persistence.sequence.impl.query.GetSequenceChannelsBySequenceIdQuery
import io.sparkled.persistence.sequence.impl.query.GetSongAudioBySequenceIdQuery
import io.sparkled.persistence.sequence.impl.query.GetStageBySequenceIdQuery
import io.sparkled.persistence.sequence.impl.query.SaveRenderedStagePropsQuery
import io.sparkled.persistence.sequence.impl.query.SaveSequenceChannelsQuery
import io.sparkled.persistence.sequence.impl.query.SaveSequenceQuery
import java.util.UUID
import javax.inject.Singleton

@Singleton
class SequencePersistenceServiceImpl(
    private val queryFactory: QueryFactory,
    private val objectMapper: ObjectMapper
) : SequencePersistenceService {

    override fun getAllSequences(): List<Sequence> {
        return GetAllSequencesQuery().perform(queryFactory)
    }

    override fun getSequenceById(sequenceId: Int): Sequence? {
        return GetSequenceByIdQuery(sequenceId).perform(queryFactory)
    }

    override fun getStageBySequenceId(sequenceId: Int): Stage? {
        return GetStageBySequenceIdQuery(sequenceId).perform(queryFactory)
    }

    override fun getSongAudioBySequenceId(sequenceId: Int): SongAudio? {
        return GetSongAudioBySequenceIdQuery(sequenceId).perform(queryFactory)
    }

    override fun getSequenceChannelsBySequenceId(sequenceId: Int): List<SequenceChannel> {
        return GetSequenceChannelsBySequenceIdQuery(sequenceId).perform(queryFactory)
    }

    override fun getSequenceChannelByUuid(sequenceId: Int, uuid: UUID): SequenceChannel? {
        return GetSequenceChannelByUuidQuery(sequenceId, uuid).perform(queryFactory)
    }

    override fun getRenderedStagePropsBySequenceAndSong(sequence: Sequence, song: Song): RenderedStagePropDataMap {
        return GetRenderedStagePropsBySequenceQuery(sequence, song).perform(queryFactory)
    }

    override fun saveSequence(sequence: Sequence, sequenceChannels: List<SequenceChannel>): Sequence {
        val savedSequence = SaveSequenceQuery(sequence).perform(queryFactory)
        SaveSequenceChannelsQuery(savedSequence, sequenceChannels, objectMapper).perform(queryFactory)
        return savedSequence
    }

    override fun publishSequence(sequence: Sequence, channels: List<SequenceChannel>, data: RenderedStagePropDataMap) {
        val savedSequence = SaveSequenceQuery(sequence).perform(queryFactory)
        SaveSequenceChannelsQuery(savedSequence, channels, objectMapper).perform(queryFactory)
        SaveRenderedStagePropsQuery(savedSequence, data).perform(queryFactory)
    }

    override fun deleteSequence(sequenceId: Int) {
        DeleteSequencesQuery(listOf(sequenceId)).perform(queryFactory)
    }
}
